File: examples/src/bin/multi_device.rs
// Multiple AUTD3 devices arranged side by side. Run with: cargo xtask example multi_device

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, Point3, UnitQuaternion};
use autd3_rs::{Client, ClientConfig};
use autd3_rs_link_ethercrab::EtherCrabLinkOption;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("info")),
        )
        .init();

    let geometry = Geometry::new(vec![
        Autd3::new(Point3::origin(), UnitQuaternion::identity()),
        Autd3::new(
            Point3::new(Autd3::DEVICE_WIDTH, 0.0, 0.0),
            UnitQuaternion::identity(),
        ),
    ]);

    let client = Client::open(
        &geometry,
        EtherCrabLinkOption::default(),
        ClientConfig::default(),
    )
    .await?;

    println!("devices: {}", client.num_devices());
    for (i, fw) in client.read_firmware_version().await?.iter().enumerate() {
        println!("device[{i}] firmware version: {fw}");
    }

    let center = geometry.center();
    println!(
        "array center: ({:.2}, {:.2}, {:.2}) mm",
        center.x, center.y, center.z
    );

    client.close().await?;
    Ok(())
}

File: bindings/python/examples/multi_device.py
"""
Multiple AUTD3 devices arranged side by side.

Run with: cargo xtask py example multi_device
"""

import asyncio

import autd3
import autd3_link_echocat as echocat
from scipy.spatial.transform import Rotation


async def main() -> None:
    geometry = autd3.geometry.Geometry(
        [
            autd3.geometry.Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0]),
            autd3.geometry.Autd3(
                origin=(autd3.geometry.Autd3.DEVICE_WIDTH, 0.0, 0.0),
                rotation=Rotation.identity(),
            ),
        ]
    )

    client = await autd3.Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        autd3.ClientConfig(),
    )

    print("devices:", client.num_devices())
    for i, fw in enumerate(await client.read_firmware_version()):
        print(f"device[{i}] firmware version: {fw}")

    center = geometry.center()
    print(f"array center: ({center[0]:.2f}, {center[1]:.2f}, {center[2]:.2f}) mm")

    await client.close()


if __name__ == "__main__":
    asyncio.run(main())
